Acas 142

brianbolt opened the pull request:
This ticket is just for registration of molecules using RDkit, This does not yet cover the case of Similarity or Substructure searches which may require some refactoring of this code to implement. This code adds the ability to store the RDkit structure hashes for both Parent and Salts to facility the registration workflow.

bffrost commented on src/main/resources/db/migration/rdkit/postgres/V2.3.0.4__rdkit_structure.sql:
```diff
@@ -0,0 +1,17 @@
+CREATE INDEX rdkit_structure_reg_idx ON rdkit_structure(reg);
+
+
+-- LOT - Original structure/As drawn
```
@brianbolt I think these were just notes you were taking on a call of ours. Please remove them from this migration.

brianbolt replied:
I'm actually going to delete this entire migration.  Looks like hibernate will generate this table for us.  It will also create the indexes if we give it the right annotation.

brianbolt replied:
Fixed

bffrost commented on src/main/java/com/labsynch/labseer/chemclasses/rdkit/ChemStructureServiceRDKitImpl.java:
```diff
@@ -188,15 +220,44 @@ public boolean createJChemTable(String tableName, boolean tautomerDupe) {
+			if (structureTable.equalsIgnoreCase("Parent_Structure")){
+				rdkitStructure.persist();
+			} else if (structureTable.equalsIgnoreCase("SaltForm_Structure")){
+
```
What happened to SaltForm structures?

brianbolt replied:
Thanks, I added this now.

bffrost commented on src/main/java/com/labsynch/labseer/chemclasses/rdkit/ChemStructureServiceRDKitImpl.java:
```diff
@@ -117,67 +127,89 @@ public boolean dropJChemTable(String tableName) {
+		//Create empty list
+		List<RDKitStructure> rdkitStructures = new ArrayList<RDKitStructure>();
+		if (searchType.equalsIgnoreCase("FULL_TAUTOMER")){
+			rdkitStructures = RDKitStructure.findRDKitStructuresByRegEquals(serviceRDKitStructure.getReg()).getResultList();
```
From reading through the CmpdReg FE code and https://github.com/mcneilco/acas-roo-server/blob/release/1.13.6/src/main/java/com/labsynch/labseer/service/RegSearchServiceImpl.java#L76 I think this is the pre-reg search.
This should be doing `findRDKitStructuresByPreRegEquals(serviceRDKitStructure.getPreReg())`

brianbolt replied:
Thanks. Makes sense.

brianbolt replied:
I fixed this.

bffrost reviewed (changes requested):
Largely makes sense. Thanks for the refactor of `getRDKitStructureFromService`. Noted a few missing pieces.

brianbolt commented on src/main/java/com/labsynch/labseer/domain/RDKitSaltStructure.java:
```diff
@@ -0,0 +1,14 @@
+@RooToString
+@RooJson
+@RooJpaActiveRecord(sequenceName = "RDKIT_STRUCTURE_PKSEQ", finders = { "findRDKitSaltStructuresByRegEquals"})
+public class RDKitSaltStructure extends RDKitStructure {
```
They won't share table.  Hibernate/JPA/Roo automatically create a different table (was a weird pull request because I had a sql migration which I don't actually need).

brianbolt replied:
I added the rkit_salt_structure migration for the rdkit structure subclass for salts

brianbolt commented on src/main/java/com/labsynch/labseer/domain/RDKitStructure.java:
```diff
@@ -0,0 +1,53 @@
+
+    @NotNull
+    @DateTimeFormat(style="M-")
+    private Date recordedDate;
```
Added it.

brianbolt commented:
@bffrost Hold off on review realized this morning I missed fixing/adding the salt form structure table and logic per your comment.

brianbolt commented on src/main/java/com/labsynch/labseer/domain/RDKitStructure.java:
```diff
@@ -0,0 +1,53 @@
+@RooJavaBean
+@RooToString
+@RooJson
+@RooJpaActiveRecord(sequenceName = "RDKIT_STRUCTURE_PKSEQ", inheritanceType = "TABLE_PER_CLASS", finders = { "findRDKitStructuresByRegEquals"})
```
I added this.

brianbolt commented:
@bffrost I just pushed a bunch of changes to this branch.

1.  I fixed the code per the PR comments
2. I fixed the validation mode by adding and RDkitDryRunStructure table (validation loads this table with structures, then does dupe checks against it during validation. This update also required a fix a few other random bugs.
3.  As part of 2. I got rid of a bunch of the ChemStructureService required methods around "jchem" tables.  They fell into 2 categories:
- Completely unused (`deleteAllJChemTableRows`, `deleteJChemTableRows`, `createJchemPropertyTable`)
- Used but now simplified (`dropJChemTable`, `createJChemTable`).  These 2 methods were always used in conjunction so I combined them into a `truncateStructureTableMethod`.

bffrost commented on src/main/java/com/labsynch/labseer/service/ChemStructureService.java:
```diff
@@ -26,12 +26,10 @@
 			String plainTable, String searchType) throws CmpdRegMolFormatException;
 
-	public boolean createJChemTable(String tableName, boolean tautomerDupe);
+	public boolean truncateStructureTable(String structureTable);
```
That's great to remove all these mentions of JChem and combine them here. I was worried these functions were used in the flyway migrations that create JChem tables but those seem to have their own functions defined locally: https://github.com/mcneilco/acas-roo-server/blob/master/src/main/java/com/labsynch/labseer/db/migration/postgres/jchem/V2_0_0_7__Create_Jchem_Tables.java

bffrost commented on src/main/java/com/labsynch/labseer/chemclasses/rdkit/ChemStructureServiceRDKitImpl.java:
```diff
@@ -291,23 +406,50 @@ public double getExactMass(String molStructure) throws CmpdRegMolFormatException
+
+			//Create empty list
+			List<RDKitStructure> rdkitStructures = new ArrayList<RDKitStructure>();
+			if (searchType.equalsIgnoreCase("FULL_TAUTOMER")){
```
Is there a reason both this and `searchMolStructures` contain search implementations? Not inherently a problem, just looking at how many different `search` functions there are, and wondering if we can refactor it down to only one function that actually does searching, and keep the rest as wrappers.

I do see that this and `searchMolStructures` return different result types: `CmpdRegMolecule[]` vs `int[]` but it still seems that could be wrapped.

brianbolt replied:
I think this makes sense to refactor now. I will do that.

brianbolt replied:
Just merged the refactor.

bffrost reviewed (approved):
This is a lot to consume so not 100% confident I caught everything, but looks good as far as I saw.

Left some optional feedback about doing slightly more refactoring within ChemStructureServiceRDKitImpl.java, but this is good to merge as-is. Nice work.
